### ISSUE TITLE
Sending messages at set times

### ISSUE THREAD

**Issue by joekroese** [grammyjs/grammY#323]
The examples in the (brilliant) documentation seem to all involve a bot that sends messages only when responding to user input. I am trying to do something different: using grammy to send messages as reminders. I.e. waiting until a set time and then sending a message.

In the [awesome grammY](https://github.com/grammyjs/awesome-grammY) repository I have found three relevant bots doing this:
1. [darvesh/remindmebot](https://github.com/darvesh/remindmebot) - Set reminder to a message in a group.
1. [vrumger/GibHugBot](https://github.com/vrumger/GibHugBot) - Get notified about GitHub events right in Telegram.
1. [Satont/twitch-notifier](https://github.com/Satont/twitch-notifier) - Get notified when a Twitch broadcast starts, ends or updates.

My use case is for [experience sampling](https://en.wikipedia.org/wiki/Experience_sampling_method), sending notifications to someone at random points in the day for psychology research or self-tracking. I need the bot to send messages at ~4 times in the day.

I'm wondering: is there a way to send 'reminder'-style messages without using an external database? The three examples use MongoDB or PostgreSQL which seems like overkill for my case with so few reminders! I am hoping that I can instead achieve this with the grammY `sessions` plugin to keep things simple.

**Comment by KnorpelSenf** [comment]
Telegram doesn't allow bots to send scheduled messages the same way users do. This means you'll have to actually call `sendMessage` when you want the message to be sent.

If you want to do things reliably, you'll have to use a database or an external service provider (https://cloud.google.com/scheduler and competitors) to send messages.

The session plugin won't help here, as it attaches a data store to each chat, and if you were to access this data from outside, you'll have a data race if a message for that chat comes in at the same time.

If you don't care too much about fault tolerance, you can use `setTimeout` and be done in a few minutes. Just don't restart your instance then :)

**Comment by joekroese** [comment]
>  The session plugin won't help here, as it attaches a data store to each chat, and if you were to access this data from outside, you'll have a data race if a message for that chat comes in at the same time.

Would you mind expanding on what you mean by "access this data from outside"? I am planning on setting the scheduled times for each chat individually (chat A will have different times to chat B) so if this idea of outside is linked to synchronisation across chats, maybe this problem of a data race is avoidable.

> If you don't care too much about fault tolerance, you can use `setTimeout` and be done in a few minutes. Just don't restart your instance then :)

I don't care much about fault tolerance! (At least at this stage of writing the software.) Are you able to share a sketch of what this approach would look like? [darvesh/remindmebot](https://github.com/darvesh/remindmebot) uses setInterval to check every 5 minutes but since I know the order of reminders, I'm hoping this will be unnecessary.

**Comment by KnorpelSenf** [comment]
> > The session plugin won't help here, as it attaches a data store to each chat, and if you were to access this data from outside, you'll have a data race if a message for that chat comes in at the same time.
> 
> Would you mind expanding on what you mean by "access this data from outside"? I am planning on setting the scheduled times for each chat individually (chat A will have different times to chat B) so if this idea of outside is linked to synchronisation across chats, maybe this problem of a data race is avoidable.

Whenever an update for a particular chat arrives from Telegram, the session data will be read, all handlers will be run, and the session data will be written back.

Updates are processed sequentially per chat, so there will never be any data races. However, if you now set a timer and then modify the session data in its callback, you're potentially running into WAR hazards.

There is no proper way to solve this problem since sessions do not support locking. The best way to mitigate this is to not use sessions but to use a database instead. That way, you'll be able to perform transactions, which guarantee data integrity.

> > If you don't care too much about fault tolerance, you can use `setTimeout` and be done in a few minutes. Just don't restart your instance then :)
> 
> I don't care much about fault tolerance! (At least at this stage of writing the software.) Are you able to share a sketch of what this approach would look like? [darvesh/remindmebot](https://github.com/darvesh/remindmebot) uses setInterval to check every 5 minutes but since I know the order of reminders, I'm hoping this will be unnecessary.

Basically, whenever you want to schedule a new task, just run `setTimeout`. That way, messages will be sent at the specified times. It's a very trivial thing to do, but you should note that stopping your process will make you lose all scheduled reminders. (That is, unless you use a database as described above.)

**Comment by joekroese** [comment]
Thanks for the explanation! That's very helpful. So it sounds like I might need to bite the bullet and abandon sessions and instead use a database.

On the second thread, I think what I'm trying to do is slightly different to the very trivial approach. My understanding is that the trivial option relies on a 1-1 relation from 'user message' to 'running a `setTimeout`' e.g.:
```js
bot.command("remind_in_five_seconds", (ctx) => {
  setTimeout(() => console.log("reminder!"), 5 * 1000);
});
```

Instead, (I think!) I'm aiming to use a recursive `setTimeout`. What I'm trying at the moment is:

```js
// just the relevant parts

bot.command("experience_sampling", (ctx) => {
  respondThenWait(ctx);
});

const now = new Date();
const inTenSeconds = new Date(now.getTime() + 10 * 1000);
const inOneMinute = new Date(now.getTime() + 60 * 1000);
const inFiveMinutes = new Date(now.getTime() + 5 * 60 * 1000);
let prompts = [now, inTenSeconds, inOneMinute, inFiveMinutes];

let counter = 0;
function respondThenWait(ctx: Context) {
  console.log(`responding to prompt #${counter}`);
  if (ctx?.msg?.chat.id != undefined) { // this returns true only on the first iteration
    bot.api.sendMessage(ctx.msg.chat.id, `responding to prompt #${counter}`);
  }
  let wait = calculateWait(ctx);
  console.log(`waiting ${wait}`);
  setTimeout((ctx) => respondThenWait(ctx), wait);
  counter++;
}
```

This works the first iteration (user receives `responding to prompt #0`), but then for all following iterations `ctx?.msg?.chat.id` is undefined and so user never receives `responding to prompt #1` etc. (It does continue sending the `console.log` for the other prompts). Is there an obvious misunderstanding of how `Context` works in grammy here? Why isn't it passed on recursively?

**Comment by KnorpelSenf** [comment]
Yes, this likely makes the most sense.

I assume you were able to fix the problem you described earlier and then removed while editing?

**Comment by joekroese** [comment]
Yes, I fixed it the morning after (a night's sleep etc. etc.) Thanks for your help!

If anyone is looking to send reminders/scheduled messages without a database, the basic patterns I used are:

1) 1 message -> 1 reminder

```js
bot.command("remind_in_five_seconds", (ctx) => {
  setTimeout(() => console.log("reminder!"), 5 * 1000);
});
```

2) 1 message -> infinite reminders, (using sessions but `ctx?.session.reminders` could be simply `reminders`)

```js
function waitThenRespond(ctx: MyContext) {
  let next_prompt = findNextReminder(ctx?.session.reminders);
  let wait = calculateWait(next_reminder);
  console.log(`waiting ${wait} until ${JSON.stringify(next_reminder?.time)}`);
  setTimeout(() => {
    respond(ctx);
    waitThenRespond(ctx);
  }, wait);
}
```

**Comment by seenyat** [comment]
I'm sorry if I'm missing something, but on [telegram API docs ](https://core.telegram.org/api/scheduled-messages)there is a parameter `schedule_date` for sendMessage, that allows you to set a date of publishing in the future. In one of the examples they even use it with bot to send scheduled message. Can't grammy use this?
